hudi-utilities/src/main/java/org/apache/hudi/utilities/keygen/MultiFormatTimestampBasedKeyGenerator.java

```diff
@@ -0,0 +1,183 @@
+    } else {
+      this.outputDateTimeZone = null;
+    }
+
+    if (timestampType == TimestampType.DATE_STRING || timestampType == TimestampType.MIXED) {
+      DataSourceUtils.checkRequiredProperties(config,
+          Collections.singletonList(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP));
+    }
+  }
+
+  private DateTimeFormatter getInputDateFormatter() {
+    if (this.configInputDateFormatList.isEmpty()) {
+      throw new IllegalArgumentException(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP + " configuration is required");
+    }
+
+    DateTimeFormatter formatter = new DateTimeFormatterBuilder()
+            .append(
+                    null,
+                    Arrays.asList(
+                            this.configInputDateFormatList.split(this.configInputDateFormatDelimiter))
+                            .stream()
+                            .map(String::trim)
+                            .map(DateTimeFormat::forPattern)
+                            .map(DateTimeFormatter::getParser)
+                            .toArray(DateTimeParser[]::new))
+            .toFormatter();
+    if (this.inputDateTimeZone != null) {
+      formatter = formatter.withZone(this.inputDateTimeZone);
+    } else {
+      formatter = formatter.withOffsetParsed();
+    }
+
+    return formatter;
+  }
+
+  @Override
+  public HoodieKey getKey(GenericRecord record) {
+    Object partitionVal = DataSourceUtils.getNestedFieldVal(record, partitionPathField, true);
+    if (partitionVal == null) {
+      partitionVal = 1L;
+    }
+
+    DateTimeFormatter inputFormatter = this.getInputDateFormatter();
+    DateTimeFormatter partitionFormatter = DateTimeFormat.forPattern(outputDateFormat);
+    if (this.outputDateTimeZone != null) {
+      partitionFormatter = partitionFormatter.withZone(outputDateTimeZone);
+    }
+
+    try {
+      long unixTime;
+      if (partitionVal instanceof Double) {
+        unixTime = ((Double) partitionVal).longValue();
+      } else if (partitionVal instanceof Float) {
+        unixTime = ((Float) partitionVal).longValue();
+      } else if (partitionVal instanceof Long) {
+        unixTime = (Long) partitionVal;
+      } else if (partitionVal instanceof CharSequence) {
+        DateTime parsedDateTime = inputFormatter.parseDateTime(partitionVal.toString());
+
+        if (this.outputDateTimeZone == null) {
+          // Use the timezone that came off the date that was passed in, if it had one
+          partitionFormatter = partitionFormatter.withZone(parsedDateTime.getZone());
+        }
+
+        unixTime = inputFormatter.parseDateTime(partitionVal.toString()).getMillis() / 1000;
+      } else {
+        throw new HoodieNotSupportedException(
+          "Unexpected type for partition field: " + partitionVal.getClass().getName());
+      }
+      DateTime timestamp = this.timestampType == TimestampType.EPOCHMILLISECONDS ? new DateTime(unixTime, outputDateTimeZone) : new DateTime(unixTime * 1000, outputDateTimeZone);
```

Comment:
There is another PR which changes the millisecond handling more gracefully..

Reply:
Cool..this was basically copy n' paste from the other one.  Would be happy to bring in the better handling.

Reply:
I looked for the PR but wasn't sure which one it was.  Might be better to get this into the branch where that change has been merged in?

Reply:
I guess Vinoth is talking about this one - https://github.com/apache/incubator-hudi/pull/1541.